modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java

```diff
@@ -997,47 +1095,59 @@ private class HostContainer implements HostInterface {
-            stagePeer = embeddedStage;
-            if (stagePeer == null) {
-                return;
+            synchronized(LOCK) {
+                stagePeer = embeddedStage;
+                if (stagePeer == null) {
+                    return;
+                }
             }
-            if (pWidth > 0 && pHeight > 0) {
-                stagePeer.setSize(pWidth, pHeight);
+            int lWidth = getCompWidth();
+            int lHeight = getCompHeight();
+            if (lWidth > 0 && lHeight > 0) {
+                embeddedStage.setSize(lWidth, lHeight);
             }
             invokeOnClientEDT(() -> {
-                if (stagePeer != null && JFXPanel.this.isFocusOwner()) {
-                    stagePeer.setFocused(true, AbstractEvents.FOCUSEVENT_ACTIVATED);
+                if (JFXPanel.this.isFocusOwner()) {
+                    embeddedStage.setFocused(true, AbstractEvents.FOCUSEVENT_ACTIVATED);
                 }
             });
             sendMoveEventToFX();
         }
 
         @Override
         public void setEmbeddedScene(EmbeddedSceneInterface embeddedScene) {
-            if (scenePeer == embeddedScene) {
-                return;
+            synchronized(LOCK) {
+                if (scenePeer == embeddedScene) {
+                    return;
+                }
+                scenePeer = embeddedScene;
             }
-            scenePeer = embeddedScene;
-            if (scenePeer == null) {
+            if (embeddedScene == null) {
                 invokeOnClientEDT(() -> {
-                    if (dnd != null) {
-                        dnd.removeNotify();
-                        dnd = null;
+                    SwingDnD lDnD = getDnD();
+                    if (lDnD != null) {
+                        lDnD.removeNotify();
+                        lDnD = null;
                     }
+                    setDnD(lDnD);
                 });
                 return;
             }
-            if (pWidth > 0 && pHeight > 0) {
-                scenePeer.setSize(pWidth, pHeight);
+            int lWidth = getCompWidth();
+            int lHeight = getCompHeight();
+            double lScaleFactorX = getScaleFactorX();
+            double lScaleFactorY = getScaleFactorY();
+            if (lWidth > 0 && lHeight > 0) {
+                embeddedScene.setSize(lWidth, lHeight);
             }
-            scenePeer.setPixelScaleFactors((float) scaleFactorX, (float) scaleFactorY);
+            embeddedScene.setPixelScaleFactors((float) lScaleFactorX, (float) lScaleFactorY);
 
             invokeOnClientEDT(() -> {
-                dnd = new SwingDnD(JFXPanel.this, scenePeer);
-                dnd.addNotify();
-                if (scenePeer != null) {
-                    scenePeer.setDragStartListener(dnd.getDragStartListener());
-                }
+                SwingDnD lDnD = getDnD();
+                lDnD = new SwingDnD(JFXPanel.this, embeddedScene);
```

Comment:
is there a need for getting getDnd() on L1146 and immediately overwriting it on L1147?
should it be
`SwingDnD lDnD = new SwingDnD(JFXPanel.this, embeddedScene);`
?